### PR TITLE
Add coveo-for-sitecore-tools to build Coveo for Sitecore assets

### PR DESCRIPTION
**Pull request by sjeanCoveo**
# Pull Request Checklist

## Image Details

- [ ] I have added NEW image tags
- [ ] I have made significant changes to existing images

## Pull Request Details

**What this PR does / why we need it**:

- Adds coveo-for-sitecore-tools to allow building Coveo for Sitecore assets in Sitecore 10+. Using those assets, the clients will be able to launch a Sitecore instance in which Coveo will already be installed. They'll also be able to select the Coveo for Sitecore version they want. 

**Which issue(s) this PR fixes**:

- None

**Special notes for your reviewer**:

- This PR contains scripts to setup the required files and folders to build Coveo for Sitecore assets. 
- It's designed to be compatible with Sitecore 10.0.0 or newer.
- No images were added or modified, everything happens locally when launching the Build.ps1 script with the 
`-CoveoVersion` parameter.
- A section has been added in the INSTRUCTIONS.md file to explain how to use the parameter.

## Pre-submit Checklist

- [x] I have read and am familiar with the contents of [CONTRIBUTING.md](https://github.com/Sitecore/docker-images/blob/master/CONTRIBUTING.md)
- [x] I have built the images locally (all relevant OS versions, including Linux if applicable)

  - Include the `.\build.ps1` command used (don't forget to exclude your credentials!)
```
  .\Build.ps1 -SitecoreVersion @("10.0.0","10.0.1","10.1.0") -Topology xp0 -IncludeModuleAssets -IncludeSxa -CoveoVersion 5.0.971.3
```
  - Ensure the build was done on a clean system (`docker system prune -af`)

- [x] I have added or updated the necessary docker-compose file(s) in the `build/windows/tests` folder
- [ ] I have updated the documentation using `build\update-documentation.ps1`  (N/A)
- [x] I have updated the `build\CHANGELOG.md` with details about the change(s) included

### CODE REVIEW

**Review by jeanfrancoislarente — COMMENTED** [review]
Before I go through the review, could you please rebase your commits (squash to a single commit)?

https://github.com/Sitecore/docker-images/blob/master/CONTRIBUTING.md

Thanks!

**Review by jeanfrancoislarente — COMMENTED** [review]
Tried to run and ran into an error:

```
.\Build.ps1 -SitecoreVersion @("10.1.0") -Topology xp0 -IncludeModuleAssets -IncludeSxa -CoveoVersion 5.0.971.3
Adjusting setup to include Coveo image...
SitecoreVersion=10.1.0 CoveoVersion=5.0.971.3 IncludeSxa=True
Coveo for Sitecore will be included in the build
05:58:02:122: Excluding experimental images.
05:58:02:544: The following images will be built:
community/modules/custom-coveo509713-assets:10.1.0-1809
community/modules/custom-coveo509713-sxa-assets:10.1.0-1809
community/modules/custom-sitecore-docker-tools-assets:10.1.0-1809
community/sitecore-xp0-custom-sxa-cm:10.1.0-ltsc2019
community/sitecore-xp0-custom-sxa-mssql:10.1.0-ltsc2019
community/sitecore-xp0-custom-sxa-solr-init:10.1.0-ltsc2019
community/sitecore-xp0-custom-sxa-sps-mssql:10.1.0-ltsc2019
05:58:06:594: Downloading 'https://static.cloud.coveo.com/coveoforsitecore/packages/v5.0.971.3/Coveo%20for%20Sitecore%20SXA%2010.1%205.0.971.3.zip' to 'C:\projects\docker-images\build\packages\Coveo for Sitecore SXA 10.1.0 5.0.971.3.scwdp.zip'...
Download complete
05:58:07:178: Downloading 'https://static.cloud.coveo.com/coveoforsitecore/packages/v5.0.971.3./Coveo%20for%20Sitecore%2010.0%205.0.971.3.scwdp.zip' to 'C:\projects\docker-images\build\packages\Coveo for Sitecore 10.1.0 5.0.971.3.scwdp.zip'...
Invoke-FileDownload: C:\projects\docker-images\build\modules\SitecoreImageBuilder\1.0.0\Public\Invoke-PackageRestore.ps1:96
Line |
  96 |                  Invoke-FileDownload -Url $fileUrl -Path $filePath
     |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | The variable '$fileStream' cannot be retrieved because it has not been set.
```

File: Build.ps1 [hunk 1]
```diff
@@ -69,6 +72,43 @@ if ([string]::IsNullOrEmpty($InstallSourcePath)) {
...
+    }
+
+    Write-Host "Coveo for Sitecore will be included in the build"
+}
+
+$ErrorActionPreference = "STOP"
+$ProgressPreference = "SilentlyContinue"
+
+# Setup the necessary folders and files to include Coveo for Sitecore Image
+if ($CoveoVersion -ne "") {
```
[the remark this answers is not in the document]
Comment (sjeanCoveo) [reply Build.ps1:96]: It shouldn't be there twice, something went wrong with the rebase

**Comment by sjeanCoveo** [comment]
@jeanfrancoislarente I asked two of my colleagues to run the scripts on their machine and none of them have been able to reproduce the error you had. I also tried to reproduce it and did not succeed. I'm quite puzzled to be honest, would you mind pulling the fork again and give it another try?

**Comment by jeanfrancoislarente** [comment]
> @jeanfrancoislarente I asked two of my colleagues to run the build on their machine and none of them have been able to reproduce the error you had. I also tried to reproduce it and did not succeed. I'm quite puzzled to be honest, would you mind pulling the fork again and give it another try?

any chance Coveo folks feel like hosting the scwdp? Otherwise people trying to build on their machines would need to set up the Sitecore Azure Toolkit dependencies? 

I did get past the other issue, I was trying to build in PowerShell 7.1.5. I tried 5.1 and it works. I just came across the SAT dependency issue now:

```
Create SAT directory C:\Projects\docker-images\build\tools\sat
Import Sitecore Azure Toolkit
Convert to WDP C:\Projects\docker-images\build\packages\Coveo for Sitecore SXA 10.0.0 10.0.1 10.1.0 5.0.971.3.zip
Microsoft.Web.Deployment.DeploymentDetailedFatalException: The SQL provider cannot run with dacpac option because of a missing dependency. Please make sure that DacFx is installed.  Learn more at: https://go.microsoft.com/fwlink/?LinkId=221672#ERROR_DACFX_NEEDED_FOR_SQL_PROVIDER.
   at Microsoft.Web.Deployment.DacHelper.ThrowIfDACCannotBeLoaded()
   at Microsoft.Web.Deployment.SqlDacPacProvider.GetAttributes(DeploymentAddAttributeContext addContext)
   at Microsoft.Web.Deployment.DeploymentObject.EnsureValidAttributes()
   at Microsoft.Web.Deployment.ManifestProvider.GetAttributes(DeploymentAddAttributeContext addContext)
   at Microsoft.Web.Deployment.DeploymentObject.EnsureValidAttributes()
   at Microsoft.Web.Deployment.DeploymentObject.get_IsValid()
   at Microsoft.Web.Deployment.DeploymentManager.CreateObjectPrivate(DeploymentProviderContext providerContext, DeploymentBaseOptions baseOptions, DeploymentObject sourceObject, String serverVersion)
   at Microsoft.Web.Deployment.DeploymentManager.CreateObject(DeploymentProviderOptions providerOptions, DeploymentBaseOptions baseOptions)
   at Microsoft.Web.Deployment.DeploymentManager.CreateObject(DeploymentWellKnownProvider provider, String path)
   at Sitecore.Cloud.Packaging.WebDeployPackages.WebDeployPackageBuilder.MSDeploySyncFromManifest(String manifestXmlPath, String tempPackage)
   at Sitecore.Cloud.Packaging.WebDeployPackages.WebDeployPackageBuilder.Build(SitecoreModuleFolderTree smTree, DirPath outputDir, String sourceName, String[] disableDacPacOptions, Boolean force)
   at Sitecore.Cloud.Cmdlets.Packaging.ConvertToSCModuleWebDeployPackage.ProcessRecord()
  ```

I'm also seeing an odd issue during the download output:

```
12:54:11:281: Downloading 'https://static.cloud.coveo.com/coveoforsitecore/packages/v5.0.971.3/Coveo%20for%20Sitecore%20SXA%2010.1%205.0.971.3.zip' to 'C:\Projects\docker-images\build\packages\Coveo for Sitecore SXA 10.1.0 5.0.971.3.scwdp.zip'
```

- Says the source is .zip but the destination is .scwdp.zip (on the download, not convert). After the package download step, I have unconverted scwdp.zip files (which contain a package.zip folder)
```
Get-ChildItem -Filter coveo*


    Directory: C:\Projects\docker-images\build\packages


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----        2021-11-08   1:02 PM       28102135 Coveo for Sitecore 10.0.0 5.0.971.3.scwdp.zip
-a----        2021-11-08   1:02 PM       28102135 Coveo for Sitecore 10.0.1 5.0.971.3.scwdp.zip
-a----        2021-11-08   1:02 PM       28102135 Coveo for Sitecore 10.1.0 5.0.971.3.scwdp.zip
-a----        2021-11-08   1:02 PM         207194 Coveo for Sitecore SXA 10.0.0 10.0.1 10.1.0 5.0.971.3.zip
-a----        2021-11-08   1:02 PM         207194 Coveo for Sitecore SXA 10.0.0 5.0.971.3.scwdp.zip
-a----        2021-11-08   1:02 PM         207194 Coveo for Sitecore SXA 10.0.1 5.0.971.3.scwdp.zip
-a----        2021-11-08   1:02 PM         207194 Coveo for Sitecore SXA 10.1.0 5.0.971.3.scwdp.zip

```
- When specifying 10.1.0 only for a build, it still downloads 10.0.0 and 10.0.1 Coveo packages.

**Comment by sjeanCoveo** [comment]
> > @jeanfrancoislarente I asked two of my colleagues to run the build on their machine and none of them have been able to reproduce the error you had. I also tried to reproduce it and did not succeed. I'm quite puzzled to be honest, would you mind pulling the fork again and give it another try?
> 
> any chance Coveo folks feel like hosting the scwdp? Otherwise people trying to build on their machines would need to set up the Sitecore Azure Toolkit dependencies?
> 
> I did get past the other issue, I was trying to build in PowerShell 7.1.5. I tried 5.1 and it works. I just came across the SAT dependency issue now:
> 
> ```
> Create SAT directory C:\Projects\docker-images\build\tools\sat
> Import Sitecore Azure Toolkit
> Convert to WDP C:\Projects\docker-images\build\packages\Coveo for Sitecore SXA 10.0.0 10.0.1 10.1.0 5.0.971.3.zip
> Microsoft.Web.Deployment.DeploymentDetailedFatalException: The SQL provider cannot run with dacpac option because of a missing dependency. Please make sure that DacFx is installed.  Learn more at: https://go.microsoft.com/fwlink/?LinkId=221672#ERROR_DACFX_NEEDED_FOR_SQL_PROVIDER.
>    at Microsoft.Web.Deployment.DacHelper.ThrowIfDACCannotBeLoaded()
>    at Microsoft.Web.Deployment.SqlDacPacProvider.GetAttributes(DeploymentAddAttributeContext addContext)
>    at Microsoft.Web.Deployment.DeploymentObject.EnsureValidAttributes()
>    at Microsoft.Web.Deployment.ManifestProvider.GetAttributes(DeploymentAddAttributeContext addContext)
>    at Microsoft.Web.Deployment.DeploymentObject.EnsureValidAttributes()
>    at Microsoft.Web.Deployment.DeploymentObject.get_IsValid()
>    at Microsoft.Web.Deployment.DeploymentManager.CreateObjectPrivate(DeploymentProviderContext providerContext, DeploymentBaseOptions baseOptions, DeploymentObject sourceObject, String serverVersion)
>    at Microsoft.Web.Deployment.DeploymentManager.CreateObject(DeploymentProviderOptions providerOptions, DeploymentBaseOptions baseOptions)
>    at Microsoft.Web.Deployment.DeploymentManager.CreateObject(DeploymentWellKnownProvider provider, String path)
>    at Sitecore.Cloud.Packaging.WebDeployPackages.WebDeployPackageBuilder.MSDeploySyncFromManifest(String manifestXmlPath, String tempPackage)
>    at Sitecore.Cloud.Packaging.WebDeployPackages.WebDeployPackageBuilder.Build(SitecoreModuleFolderTree smTree, DirPath outputDir, String sourceName, String[] disableDacPacOptions, Boolean force)
>    at Sitecore.Cloud.Cmdlets.Packaging.ConvertToSCModuleWebDeployPackage.ProcessRecord()
> ```
> 
> I'm also seeing an odd issue during the download output:
> 
> ```
> 12:54:11:281: Downloading 'https://static.cloud.coveo.com/coveoforsitecore/packages/v5.0.971.3/Coveo%20for%20Sitecore%20SXA%2010.1%205.0.971.3.zip' to 'C:\Projects\docker-images\build\packages\Coveo for Sitecore SXA 10.1.0 5.0.971.3.scwdp.zip'
> ```
> 
> * Says the source is .zip but the destination is .scwdp.zip (on the download, not convert). After the package download step, I have unconverted scwdp.zip files (which contain a package.zip folder)
> 
> ```
> Get-ChildItem -Filter coveo*
> 
> 
>     Directory: C:\Projects\docker-images\build\packages
> 
> 
> Mode                 LastWriteTime         Length Name
> ----                 -------------         ------ ----
> -a----        2021-11-08   1:02 PM       28102135 Coveo for Sitecore 10.0.0 5.0.971.3.scwdp.zip
> -a----        2021-11-08   1:02 PM       28102135 Coveo for Sitecore 10.0.1 5.0.971.3.scwdp.zip
> -a----        2021-11-08   1:02 PM       28102135 Coveo for Sitecore 10.1.0 5.0.971.3.scwdp.zip
> -a----        2021-11-08   1:02 PM         207194 Coveo for Sitecore SXA 10.0.0 10.0.1 10.1.0 5.0.971.3.zip
> -a----        2021-11-08   1:02 PM         207194 Coveo for Sitecore SXA 10.0.0 5.0.971.3.scwdp.zip
> -a----        2021-11-08   1:02 PM         207194 Coveo for Sitecore SXA 10.0.1 5.0.971.3.scwdp.zip
> -a----        2021-11-08   1:02 PM         207194 Coveo for Sitecore SXA 10.1.0 5.0.971.3.scwdp.zip
> ```
> 
> * When specifying 10.1.0 only for a build, it still downloads 10.0.0 and 10.0.1 Coveo packages.

Thanks for the quick answer. I'll look into the issues you mentionned, if you included only Sitecore 10.1.0 you shouldn't have other versions included. For the scwdp for Coveo SXA it's coming soon. As mentionned in the PR the Convert-Coveo-Sxa-To-Wdp.ps1 script is temporary and so is the Sitecore Azure Toolkit dependency.

**Comment by sjeanCoveo** [comment]
@jeanfrancoislarente I investigated the issues you mentioned, here are the results:
- **SAT issue:** 
  We can't really avoid using it for now. We won't have to use it anymore after the next Coveo for Sitecore release, as the 
  scwdp will be availaible directly on Coveo's website. I can add a note in the INSTRUCTIONS.md file to mention that it may 
  require aditionnal dependencies.
- **zip saved as scwdp:** 
  fixed by https://github.com/Sitecore/docker-images/pull/703/commits/5de92ba7462495f292361b6272da7effa482126b
- **Using one version downloads the others too:**
  unable to reproduce

**Comment by michael-jensen** [comment]
Hi @sjeanCoveo with your recent [November release](https://docs.coveo.com/en/l6h90340/coveo-for-sitecore-v5/november-19-2021-release-5-0-1039-1) I am wondering if you'll be updating the details in the PR (i.e. removing the SAT dependency, etc.) soon.

We are looking forward to trying out this release in a managed cloud environment! 😄

**Comment by sjeanCoveo** [comment]
Hi @michael-jensen, I just commited the modifications , the SAT dependency was removed and  everything else was adjusted accordingly. We are now directly retrieving the .scwdp from the Coveo website. Let me know if there is any other change required.

**Comment by sjeanCoveo** [comment]
@jeanfrancoislarente the SAT dependency was removed. Is it possible to have an update on this PR?

**Comment by jeanfrancoislarente** [comment]
> @jeanfrancoislarente the SAT dependency was removed. Is it possible to have an update on this PR?

@sjeanCoveo the team is slammed this week for "a big demo" (aren't they all?) - we'll be able to look at it late this week or early next week.

**Review by jeanfrancoislarente — CHANGES_REQUESTED** [review]
Please ensure you squash commit. tags/* files should remain unmodified and/or not be listed in the changed files.

**Comment by sjeanCoveo** [comment]
Any update on this PR ?

**Comment by jeanfrancoislarente** [comment]
It has been a VERY long while @sjeanCoveo but I finally gave it a go today.

Error I'm getting now has something to do with the asset public link being expired maybe? I used the command sample in the PR instructions.

```
11:14:09:644: Downloading 'https://static.cloud.coveo.com/coveoforsitecore/packages/v5.0.971.3/Coveo%20for%20Sitecore%20SXA%2010.1%205.0.971.3.scwdp.zip' to 'C:\projects\docker-images\build\packages\Coveo for Sitecore SXA 10.1.0 5.0.971.3.scwdp.zip'...
Invoke-FileDownload: C:\projects\docker-images\build\modules\SitecoreImageBuilder\1.0.0\Public\Invoke-PackageRestore.ps1:96
Line |
  96 |                  Invoke-FileDownload -Url $fileUrl -Path $filePath
     |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | The variable '$fileStream' cannot be retrieved because it has not been set.
```

**Comment by sjeanCoveo** [comment]
> It has been a VERY long while @sjeanCoveo but I finally gave it a go today.
> 
> Error I'm getting now has something to do with the asset public link being expired maybe? I used the command sample in the PR instructions.
> 
> ```
> 11:14:09:644: Downloading 'https://static.cloud.coveo.com/coveoforsitecore/packages/v5.0.971.3/Coveo%20for%20Sitecore%20SXA%2010.1%205.0.971.3.scwdp.zip' to 'C:\projects\docker-images\build\packages\Coveo for Sitecore SXA 10.1.0 5.0.971.3.scwdp.zip'...
> Invoke-FileDownload: C:\projects\docker-images\build\modules\SitecoreImageBuilder\1.0.0\Public\Invoke-PackageRestore.ps1:96
> Line |
>   96 |                  Invoke-FileDownload -Url $fileUrl -Path $filePath
>      |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>      | The variable '$fileStream' cannot be retrieved because it has not been set.
> ```

Hi @jeanfrancoislarente , actually the version in the PR instruction is outdated, this version required SAT in order to build the scwdp. Use `5.0.1039.1` or 
 `5.0.1110.1`, for those versions the scwdp is provided directly.

**Pull request merged**